core/src/main/kotlin/org/eln2/data/ComponentGraph.kt

```diff
@@ -0,0 +1,278 @@
+package org.eln2.data
+
+import org.eln2.debug.dprintln
+import java.lang.StringBuilder
+import java.util.*
+
+open class Edge(val a: Vertex, val b: Vertex): Iterable<Vertex> {
+	override fun iterator() = arrayOf(a, b).iterator()
+}
+
+open class Vertex(val graph: ComponentGraph) {
+	var component = graph.constructComponent(this)
+	val incident: MutableMap<Vertex, Edge> = mutableMapOf()
+}
+
+open class ConnectedComponent(initialVertex: Vertex) {
+	val vertices: MutableSet<Vertex> = mutableSetOf(initialVertex)
+	
+	internal open fun add(v: Vertex) = vertices.add(v)
+	internal open fun remove(v: Vertex) = vertices.remove(v)
+	internal open fun clear() = vertices.clear()
+
+	internal var realized = false
+	internal open fun realize() {}
+	internal open fun unrealize() {}
+}
+
+interface ComponentListener {
+	fun componentRealized(c: ConnectedComponent)
+	fun componentUnrealized(c: ConnectedComponent)
+}
+
+interface VertexListener {
+	fun vertexAdded(v: Vertex)
+	fun vertexRemoved(v: Vertex)
+}
+
+interface EdgeListener {
+	fun edgeAdded(e: Edge)
+	fun edgeRemoved(e: Edge)
+}
+
+open class ComponentGraph {
+	val vertices: MutableSet<Vertex> = mutableSetOf()
+	val edges: MutableSet<Edge> = mutableSetOf()
+	var components: MutableSet<ConnectedComponent> = mutableSetOf()
+	
+	val componentListeners: MutableSet<ComponentListener> = mutableSetOf()
+	val vertexListeners: MutableSet<VertexListener> = mutableSetOf()
+	val edgeListeners: MutableSet<EdgeListener> = mutableSetOf()
+	
+	fun addListener(listener: Any) {
+		if(listener is ComponentListener) componentListeners.add(listener)
+		if(listener is VertexListener) vertexListeners.add(listener)
+		if(listener is EdgeListener) edgeListeners.add(listener)
+	}
+	
+	fun removeListener(listener: Any) {
+		componentListeners.remove(listener)
+		vertexListeners.remove(listener)
+		edgeListeners.remove(listener)
```

Comment:
Checking the type will make this faster, don't try this operation needlessly on 3 lists. That said, does the compiler optimize this? Not sure it does.